samza-core/src/main/java/org/apache/samza/table/remote/AsyncHelper.java

```diff
@@ -0,0 +1,189 @@
+ * executed on the threads of the underlying native data store client. This could be useful when
+ * application might execute long-running operations upon future completions.
+ *
+ * Each table instance should create a separate AsyncHelper instance unless they can share the
+ * same throttler instance, ie. tagged the same way.
+ *
+ * @param <K> type of the table key
+ * @param <V> type of the table record
+ */
+public class AsyncHelper<K, V> {
+  private final Throttler<K, V> throttler;
+
+  @VisibleForTesting
+  final ExecutorService tableExecutor;
+  final ExecutorService callbackExecutor;
+
+  public AsyncHelper(String tableId, int maxRequests, Throttler<K, V> throttler, ExecutorService callbackExecutor) {
+    this.throttler = throttler;
+    this.callbackExecutor = callbackExecutor;
+    tableExecutor = new ThreadPoolExecutor(
+        1, 1, 0L, TimeUnit.MILLISECONDS, // same values as Executors.newSingleThreadedExecutor
+        maxRequests == Integer.MAX_VALUE ? new LinkedBlockingQueue<>() : new ArrayBlockingQueue<>(maxRequests),
+        arg -> {
+        Thread thread = new Thread(arg);
+        thread.setName(tableId + "-async-helper");
+        thread.setDaemon(true);
+        return thread;
+      });
+  }
+
+  /**
+   * Decorate a CompletableFuture with result handling.
+   * @param ioFuture future of the table operation
+   * @param future future returned to the caller
+   * @param startNs nano time of the start time
+   * @param timer latency metric to be updated
+   * @param <T> return type
+   * @return CompletableFuture to return to the caller
+   */
+  private <T> CompletableFuture<T> decorateFuture(
+      CompletableFuture<T> ioFuture, CompletableFuture<T> future, long startNs, Timer timer) {
+    return callbackExecutor != null ?
+        ioFuture.handleAsync((result, e) -> {
+            if (e != null) {
+              future.completeExceptionally(e);
+            } else {
+              timer.update(System.nanoTime() - startNs);
+              future.complete(result);
+            }
+            return result;
+          }, callbackExecutor) :
+        ioFuture.handle((result, e) -> {
+            if (e != null) {
+              future.completeExceptionally(e);
+            } else {
+              timer.update(System.nanoTime() - startNs);
+              future.complete(result);
+            }
+            return result;
+          });
+  }
+
+  /**
+   * Execute an async request given a table key
+   * @param key key of the table record
+   * @param method method to be executed
+   * @param timer latency metric to be updated
+   * @param <T> return type
+   * @return CompletableFuture of the operation
+   */
+  public <T> CompletableFuture<T> execute(
+      K key, Function<K, CompletableFuture<T>> method, Timer timer) {
+    CompletableFuture<T> future = new CompletableFuture<>();
+    final long startNs = System.nanoTime();
+    CompletableFuture.supplyAsync(() -> {
```

Comment:
This was added when I had to consume the result for getAllAsync and do the partial result checking.  Now that we don't have the check. This can be simplified and simply return the IO future composed with our logic, eg. update timer, etc.